packages/worker/src/worker-options.ts

```diff
@@ -411,13 +416,109 @@ export function addDefaultWorkerOptions(options: WorkerOptions): WorkerOptionsWi
+    );
+    Runtime.instance().logger.warn(`This might degrade performances and could result in OOM errors.`);
+    Runtime.instance().logger.warn(
+      `For optimal performance, try setting '--max-old-space-size' between 25% and 50% of system memory.`
+    );
+  }
+
+  return {
+    maxCachedWorkflows: Math.floor(
+      Math.max((systemResources.memory + systemResources.swap - heapSizeLimit) / GiB - 1, 1) * 250
+    ),
+  };
+}
+
+/**
+ * Inspect the execution environment and return information about available system resources,
+ * taking into account applicable constraints that can be discovered.
+ * At present, it accounts for CGroups v1 and v2 constraints on memory and swap.
+ */
+export function inspectSystemResources(): { memory: number; swap: number } {
+  const systemMemory = os.totalmem();
+
+  const limits = {
+    memory: systemMemory,
+    swap: 0,
+  };
+
+  if (process.platform === 'linux') {
+    // Check for v2 style cgroups
+    if (fs.existsSync('/sys/fs/cgroup/cgroup.controllers')) {
+      // Maximum size (in bytes) of memory usage allowed by this container. 'max' if no constraint.
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - max => no constraint
+      const memoryMax = tryReadFileSync('/sys/fs/cgroup/memory.max');
+      if (isNumber(memoryMax)) limits.memory = Math.min(Number(memoryMax), limits.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size EXCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.swap.max will report 256mb
+      // Defaults to one time the size of maximum memory (if constrained), or 'max' if no constraint on memory
+      //
+      // Examples:
+      //  - 268435456 => 256 Mb
+      //  - max => no constraint
+      const swapMax = tryReadFileSync('/sys/fs/cgroup/memory.swap.max');
+      // For heuristics, don't use more than one time the RAM in swap
+      if (isNumber(swapMax)) limits.swap = Math.min(Number(swapMax), systemMemory);
+    }
+
+    // Check for v1 style cgroups: /sys/fs/cgroup/memory/memory.limit_in_bytes
+    if (fs.existsSync('/sys/fs/cgroup/memory/memory.limit_in_bytes')) {
+      // Maximum size (in bytes) of memory usage allowed by this container. '9223372036854771712' if no constraint
+      // Examples:
+      //  - 536870912 => 512 Mb
+      //  - 9223372036854771712 => no constraint
+      const memoryMax = tryReadFileSync('/sys/fs/cgroup/memory/memory.limit_in_bytes');
+      if (isNumber(memoryMax)) limits.memory = Math.min(Number(memoryMax), limits.memory);
+
+      // Maximum size (in bytes) of swap usage allowed by this container. The reported size INCLUDES the size of memory itself.
+      // For example, assuming `docker run --memory 512mb --memory-swap 768mb`, memory.swap.max will report 768mb. File might
```

Comment:
It won't report 1280mb?

Reply:
Note to myself: `s/memory.swap.max/memory.memsw.limit_in_bytes/`